easybuild/easyblocks/generic/cargo.py

```diff
@@ -392,44 +388,72 @@ def _setup_offline_config(self, git_sources):
-                    self.log.debug("Writing config.toml entry for git repo: %s branch %s, rev %s",
-                                   git_repo, branch, rev)
-                    branch_info = 'branch = "%s"\n' % branch if branch else '\n'
-                    write_file(config_toml,
-                               CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev) + branch_info,
-                               append=True)
+                git_branch = self._get_crate_git_repo_branch(src['crate_name'])
+                if git_branch:
+                    self.log.debug(f"Writing config.toml entry for git repo: {git_repo} branch {git_branch}, rev {rev}")
+                    write_file(
+                        config_toml,
+                        CONFIG_TOML_SOURCE_GIT_BRANCH.format(url=git_repo, rev=rev, branch=git_branch),
+                        append=True
+                    )
                 else:
-                    self.log.debug("Preparing config.toml entry for git repo: %s", git_repo)
-                    # Fallback for (older) easyconfigs that don't specify the branch:
-                    # Patch the whole repo, which is only possible if there is a single revision used.
-                    config_toml_git_patches[git_repo] += CONFIG_TOML_PATCH_GIT_CRATES.format(
-                        crate=src['crate_name'], path=src_dir)
+                    self.log.debug(f"Writing config.toml entry for git repo: {git_repo} , rev {rev}")
+                    write_file(
+                        config_toml,
+                        CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev),
+                        append=True
+                    )
             else:
                 self.log.debug("Writing config.toml entry for git repo: %s rev %s", git_repo, rev)
                 # Workspace sources stay in their own separate folder.
                 # We cannot have a `directory = "<dir>"` entry where a folder containing a workspace is inside
                 write_file(config_toml,
                            CONFIG_TOML_SOURCE_GIT_WORKSPACE.format(url=git_repo, rev=rev, workspace_dir=src_dir),
                            append=True)
-        for git_repo, config_crates in config_toml_git_patches.items():
-            self.log.debug("Writting config.toml patch entry for git repo: %s", git_repo)
-            write_file(config_toml, CONFIG_TOML_PATCH_GIT.format(repo=git_repo, crates=config_crates), append=True)
 
         # Use environment variable since it would also be passed along to builds triggered via python packages
         env.setvar('CARGO_NET_OFFLINE', 'true')
 
+    def _get_crate_git_repo_branch(self, crate_name):
+        """
+        Seek the dependency definition for given crate in all Cargo.toml files of sources
+        Return branch target for given crate_name if any
+        """
+        cargo_toml_glob = os.path.join(self.builddir, f'{self.name}-{self.version}', '**', 'Cargo.toml')
```

Comment:
"name-version" seems to be a bit unreliable. IIRC `self.start_dir` is exactly that:
```suggestion
        cargo_toml_glob = os.path.join(self.start_dir', '**', 'Cargo.toml')
```

I guess you can even fold this into the `glob.glob` to save the variable.

BTW: Why not use `from glob import glob? Avoids the `glob.glob` weirdness ;-)